mgmt-vrf-design-doc.md

```diff
@@ -0,0 +1,313 @@
+VRF On the other hand is a network layer 3 feature and as such should really only impact FIB lookups. While the isolation provided by a namespace includes the route tables, a network namespace will have  a significant overhead as we will end up having two network stacks.
+
+Using network name space as VRF’s will involve having services replicated across all VRF’s – for example we will have run two instances of lldp/ssh for Management and Default VRF’s respectively. 
+Also in terms of extensibility, using namespaces for Multi VRF means N-VRF’s which will have N-application instances and also the additional complexity of aggregating all the data across these instances.
+ 
+With L3-Mdev on 4.9 , VRF’s are created and enslaved to an L3 Interface and services like ssh can be shared across VRF’s thereby avoiding overhead of running multiple instances of IP services.
+
+
+Phase-1
+-------
+
+In this Phase mgmt-vrf feature will be enabled by default when SONiC boots with Debian 4.9 kernel, configuration of mgmt-vrf will be
+done using linux commands.
+
+
+DB Schema
+---------
+
+The config_db.json schema is updated to have the vrfname in MGMT_INTERFACE. The schema representation will look like below
+
+    "MGMT_INTERFACE": {
+        "eth0|10.11.150.19/24": {
+            "gwaddr": "10.11.150.254"
+            "vrfname": "mgmt-vrf"
+         }
+    }
+
+The vrfname if set to mgmt-vrf will program the linux kernel with the vrf configuration and create a mgmt-vrf for the mgmt
+port eth0. If it is set to "None" then the mgmt-vrf will not be configured on the system. Management VRF will be enabled by default when SONiC boots. 
+
+
+Flow
+-----
+
+By default, mgmt-vrf will be configured in Debian 4.9 kernel, when SONiC boots there will be isolation of traffic between
+mgmt traffic and data traffic. This is acheived by setting the vrfname to mgmt-vrf in the minigraph.py file as default setting.
+The interfaces.j2 file checks for the vrfname tag to see if it is set to mgmt-vrf. If the vrfname is set then it creates the
+mgmt-vrf configuration by executing the lniux commands. Debian 4.9 kernel does not support running services per VRF, by enabling
+the tcp_l3mdev_accept=1 the services will work across all VRF's. Below diagram shows the flows of event for creating mgmt-vrf on
+SONiC using Debian 4.9
+
+![](https://i.imgur.com/Ge6SNX4.png)
+
+
+Phase-2
+-------
+
+Later when kernel is upgraded to 4.15 or later support for running services in separate VRF will be available. New CLI command will be
+provided to create and delete VRF using config commands. Config-save command will be used to save the configuration to config_db.
+
+Proposed New CLI
+----------------
+
+The following CLI can be used to Configure and show vrf's.
+
+    config vrf add/del <vrfName>
+    config vrf member add/del <vrfName> <interfaceName>
+    show vrf config
+    show vrf brief
+    show vrf <vrfname>
+

```

Comment:
Adding some comments on behalf of the Dilip Kumar/VRF Team:


CLI Syntax in Mgmt VRF HLD | Feedback/suggestions | Notes
-- | -- | --
config vrf add mgmt | config vrf add Vrf-mgmt | VRF name should be prefixed by "Vrf"
config vrf del mgmt | config vrf del Vrf-mgmt |  
config interface ip add eth0 ip/mask gatewayIP | config interface ip add eth0 <IP/Mask> | The default GW must be configured as a regular static IP route command (ip route add 0.0.0.0/0 GW)
config interface ip remove eth0 ip/mask |   |  
  |   |  
show mgmt-vrf | show vrf Vrf-mgmt [--verbose] | This command should display the data from ConfigDB instead from kernel, like for other VRFs
show mgmt-vrf routes | show ip route vrf Vrf-mgmt | This should display routes from Zebra instead from Kernel. Behavior should be similar to other user VRFs